### ISSUE TITLE
[analyzer] offer fix to add `late` for `invalid_reference_to_this`

### ISSUE THREAD

**Issue by asashour** [dart-lang/sdk#49218]
```dart
class A {
  B b = B(this); // <- here
}

class B {
  A a;

  B(this.a);
}
```

There is an [invalid_reference_to_this](https://dart.dev/tools/diagnostic-messages?utm_source=dartdev&utm_medium=redir&utm_id=diagcode&utm_content=invalid_reference_to_this#invalid_reference_to_this).

It took me some time to figure out that `late` is what is needed, since this type of referencing is also possible in Java for example.

Would it make sense to offer a quick fix to add `late` in this case?

**Comment by eernstg** [comment]
Yes, an initializing expression for a late instance variable does have access to `this`, but when it's a non-late variable it does not have that access. However, there could also be situations where the "best" modification to the code would be to change initializer lists on constructors in the same class, because they may well be able to do the same thing, and they don't incur any dynamic checking on the initialization status of the variable during the rest of the lifetime of the object.

In the given example we have a cyclic object graph (an `A` has a reference to a `B` that has a reference back to the same `A`), and in this case there's no hope that we can use initializer lists (because they enforce an ordering on the creation of objects). But we could have done it in this variant:

```dart
class A {
  B b;
  A(int i): B(i);
}

class B {
  int i;
  B(this.i);
}
```

So if it's ok that the fix is at times not optimal then the change that adds `late` would be fine, but we put more emphasis on avoiding `late` (and the run-time costs associated with `late`) if at all possible, then it's perhaps better to emit some advice about the set of possible fixes and let the developer make a choice.

**Comment by bwilkerson** [comment]
> However, there could also be situations where the "best" modification to the code would be to change initializer lists on constructors ...

I'm having trouble coming up with a concrete example of such a case, which might indicate that marking the field as `late` in a situation like this (where `this` is being referenced in the initializer) might just be the most common appropriate fix. If that's the case, then I have no concerns about adding such a fix.

**Comment by eernstg** [comment]
Here's an example:

```dart
class A {
  int i;
  int j = i + 1; // Error, no access to `this`.
  A(this.i);
}

class AfixInitializer {
  int i;
  int j;
  AfixInitializer(this.i): j = i + 1;
}

class AfixLate {
  int i;
  late int j = i + 1;
  AfixLate(this.i);
}
```

**Comment by FMorschel** [comment]
I'll take a look at this one!

Similar to:

- https://github.com/dart-lang/sdk/issues/60729

**Comment by FMorschel** [comment]
Here is the CL https://dart-review.googlesource.com/c/sdk/+/493260
